core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1311,6 +1311,56 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * The implementation will try to preserve chunks like [[Stream.merge]].
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire stream fails and the finalizer of the
+	* inner stream runs before the outer one.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Semaphore[F2](1)).flatMap { guard =>
+      Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+        Stream.eval(SignallingRef.apply(false)).flatMap { innerFailed =>
+          def runInner(o: O, halt: Deferred[F2, Unit]) =
+            Stream.eval(guard.acquire) >> // guard inner to prevent parallel inner streams
+              f(o)
+                .interruptWhen(halt.get.attempt)
+                .handleErrorWith { e =>
+                  Stream.eval(innerFailed.set(true)) >> Stream.raiseError[F2](e)
+                }
+                // upon failure, do not release the guard(prevents the next stream from running)
+                .onFinalize(innerFailed.get.flatMap { if (_) F2.unit else guard.release })
+
+          this
+          // although failure on inner propagates, it is still possible for parJoin to
+          // keep pulling before it fails so we interrupt it anyway
+            .interruptWhen(innerFailed)
```

Comment:
I'm curious if you have seen any failing tests because of this. Even with that, it's still possible for `this` to keep pulling before being interrupted, and the mechanism for error propagation in `join` is similar to this anyway, so I'm wondering if this adds anything or not, especially given that the `Semaphore` should be guarding things regardless.

Reply:
Investigating, currently leaving ` .interruptWhen(innerFailed)` out causes tests to fail.

Reply:
It's quite possible that I'm wrong and you need that. I'm only worried that `interruptWhen` makes the failure _less likely_ rather than solving it, so I'd like to understand what's going on

Reply:
Ok, this seems more like a test issue now.  The ordering of finalisers are correct after manually inspecting the test output for several failed runs. The test was flakey because the `verdict` checking (which is also a stream) ran before the outer finaliser had a chance to record it. Working on a fix now.

Reply:
Done

Reply:
Cool. I think `innerFailed` can be a normal `Ref` now (more lightweight), and eventually it can go away entirely once `bracketCase` is in (#1269 )